### PR TITLE
ceph-fuse: restrict already_fuse_mounted function only for linux

### PR DESCRIPTION
**Pull request by nmshelke**
Signed-off-by: Nikhilkumar Shelke <nshelke@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: qa/tasks/cephfs/test_misc.py [hunk 1]
```diff
@@ -30,6 +31,9 @@ def test_statfs_on_deleted_fs(self):
         p.wait()
 
     def test_fuse_mount_on_already_mounted_path(self):
+        if platform.system() != "Linux":
+            self.skipTest("Require FUSE client")
```
Comment (lxbsz) [inline qa/tasks/cephfs/test_misc.py:35]: `s/"Require FUSE client"/"Require Linux platform"/`

**Comment by lxbsz** [comment]
@nmshelke You need to create a tracker to backport it.

File: src/mount.fuse.ceph [hunk 2]
```diff
@@ -74,7 +75,9 @@ def main(arguments):
     mount_cmd.communicate()
 
     if (mount_cmd.returncode != 0):
-        if (mount_cmd.returncode != errno.EBUSY):
+        if (mount_cmd.returncode != errno.EBUSY) and (platform.system() == "Linux")
```
Comment (kotreshhr) [inline src/mount.fuse.ceph:78]: if and else block are same ?

Comment (kotreshhr) [reply src/mount.fuse.ceph:78]: I think you should do
```
if (platform.system() == "Linux"):
    if (mount_cmd.returncode != errno.EBUSY):
        print("Mount failed with status code: {}".format(mount_cmd.returncode))
else:
    print("Mount failed with status code: {}".format(mount_cmd.returncode))
```

Comment (nmshelke) [reply src/mount.fuse.ceph]: @kotreshhr Updated as per suggestion.

**Review by lxbsz — APPROVED** [review]

**Review by kotreshhr — APPROVED** [review]

File: src/client/fuse_ll.cc [hunk 3]
```diff
@@ -242,6 +248,7 @@ static int already_fuse_mounted(const char *path, bool &already_mounted)
 
   return err;
 }
+#endif
```
Comment (vshankar) [inline src/client/fuse_ll.cc:251]: Better would be to have an `#else` block that mimics old behavior for non-linux platforms. Caller need to wrapped under `#if ...` block.

Comment (nmshelke) [reply src/client/fuse_ll.cc]: @vshankar Updated code.

File: src/client/fuse_ll.cc [hunk 4]
```diff
@@ -175,13 +180,13 @@ class CephFuse::Handle {
 
 static int already_fuse_mounted(const char *path, bool &already_mounted)
 {
+  int err = 0;
+  already_mounted = false;
+#if defined(__linux__)
```
Comment (vshankar) [inline src/client/fuse_ll.cc:185]: Well, I meant something like this:

```
#if defined(__linux__)
static int already_fuse_mounted(const char *path, bool &already_mounted) {
...
...
}
# else // non-linux platforms
static int already_fuse_mounted(const char *path, bool &already_mounted) {
...
...
}
#endif
```
Just that its a _tiny_ bit easier to read.

File: src/client/fuse_ll.cc [hunk 5]
```diff
@@ -242,6 +248,14 @@ static int already_fuse_mounted(const char *path, bool &already_mounted)
 
   return err;
 }
+#else // non-linux platforms
+static int already_fuse_mounted(const char *path, bool &already_mounted)
+{
+  int err = 0;
```
Comment (vshankar) [inline src/client/fuse_ll.cc:254]: required? just `return 0` :)

File: src/client/fuse_ll.cc (hunk 4 above)
Comment (nmshelke) [reply src/client/fuse_ll.cc]: @vshankar Done.

File: src/client/fuse_ll.cc (hunk 5 above)
Comment (nmshelke) [reply src/client/fuse_ll.cc]: Done.

**Review by vshankar — APPROVED** [review]
LGTM

**Review by wjwithagen — APPROVED** [review]
During the weekend this got build on my FreeBSD builders.
That went oke.

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Apr-14

**Pull request merged**